kernel/thread.c

```diff
@@ -190,18 +192,45 @@ thread_exit(uchar status)
 ** Current thread and current virtual address space must be locked.
 */
 status_t
-thread_exec(char const *path __unused /* TODO */)
+thread_exec(char const *path)
 {
+	status_t err;
+	uchar *exec_data;
+	void *tmp;
+	size_t size;
+	size_t r;
+	struct file_handle *file;
 	status_t s;
-	struct initrd_virt const *virt = initrd_get_virtual();
 
 	if ((s = thread_detach_and_create_vaspace())) {
 		return (s);
 	}
+	file = NULL;
+	exec_data = NULL;
+	size = 0;
+	if ((err = fs_open(path, &file))) {
```

Comment:
You can leave this for an other commit, but `path` actually belongs to the user, and user memory was deleted after `thread_detach_and_create_vaspace()`, so this pointer is invalid.

This works because you called `thred_exec` with a kernel string, but it won't work for userspace softwares.

Reply:
I thought I was missing something. So I tested, guess I got tricked :/

Reply:
Wait, wouldn't just moving the vaspace destruction below the file reading fix the problem?

Reply:
You would still need to verify that `path` is valid, but yes, it should work.